### PR TITLE
[WIP] Implement range interval partition

### PR DESCRIPTION
**Pull request by chenerlu**
This PR is try to implement range interval partition and now work on process.

### CODE REVIEW

**Comment by asfgit** [comment]
Can one of the admins verify this patch?

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder/2758/

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/carbondata-pr-spark-1.6/687/<h2>Build result: FAILURE</span></h2>[...truncated 64.18 KB...][ERROR] Re-run Maven using the -X switch to enable full debug logging.[ERROR] [ERROR] For more information about the errors and possible solutions, please read the following articles:[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoExecutionException[ERROR] [ERROR] After correcting the problems, you can resume the build with the command[ERROR]   mvn <goals> -rf :carbondata-spark-common[JENKINS] Archiving /home/jenkins/jenkins-slave/workspace/carbondata-pr-spark-1.6/processing/pom.xml to org.apache.carbondata/carbondata-processing/1.2.0-SNAPSHOT/carbondata-processing-1.2.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/jenkins-slave/workspace/carbondata-pr-spark-1.6/integration/spark-common/pom.xml to org.apache.carbondata/carbondata-spark-common/1.2.0-SNAPSHOT/carbondata-spark-common-1.2.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/jenkins-slave/workspace/carbondata-pr-spark-1.6/examples/spark/pom.xml to org.apache.carbondata/carbondata-examples-spark/1.2.0-SNAPSHOT/carbondata-examples-spark-1.2.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/jenkins-slave/workspace/carbondata-pr-spark-1.6/integration/hive/pom.xml to org.apache.carbondata/carbondata-hive/1.2.0-SNAPSHOT/carbondata-hive-1.2.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/jenkins-slave/workspace/carbondata-pr-spark-1.6/common/pom.xml to org.apache.carbondata/carbondata-common/1.2.0-SNAPSHOT/carbondata-common-1.2.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/jenkins-slave/workspace/carbondata-pr-spark-1.6/assembly/pom.xml to org.apache.carbondata/carbondata-assembly/1.2.0-SNAPSHOT/carbondata-assembly-1.2.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/jenkins-slave/workspace/carbondata-pr-spark-1.6/pom.xml to org.apache.carbondata/carbondata-parent/1.2.0-SNAPSHOT/carbondata-parent-1.2.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/jenkins-slave/workspace/carbondata-pr-spark-1.6/examples/flink/pom.xml to org.apache.carbondata/carbondata-examples-flink/1.2.0-SNAPSHOT/carbondata-examples-flink-1.2.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/jenkins-slave/workspace/carbondata-pr-spark-1.6/core/pom.xml to org.apache.carbondata/carbondata-core/1.2.0-SNAPSHOT/carbondata-core-1.2.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/jenkins-slave/workspace/carbondata-pr-spark-1.6/hadoop/pom.xml to org.apache.carbondata/carbondata-hadoop/1.2.0-SNAPSHOT/carbondata-hadoop-1.2.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/jenkins-slave/workspace/carbondata-pr-spark-1.6/integration/presto/pom.xml to org.apache.carbondata/carbondata-presto/1.2.0-SNAPSHOT/carbondata-presto-1.2.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/jenkins-slave/workspace/carbondata-pr-spark-1.6/integration/spark-common-test/pom.xml to org.apache.carbondata/carbondata-spark-common-test/1.2.0-SNAPSHOT/carbondata-spark-common-test-1.2.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/jenkins-slave/workspace/carbondata-pr-spark-1.6/integration/spark/pom.xml to org.apache.carbondata/carbondata-spark/1.2.0-SNAPSHOT/carbondata-spark-1.2.0-SNAPSHOT.pomchannel stoppedSetting status of 90bbd1a58e79bf3548e1e5cff29714ea6c928df6 to FAILURE with url https://builds.apache.org/job/carbondata-pr-spark-1.6/687/ and message: 'Tests Failed for Spark1.6 'Using context: Jenkins(Spark1.6): mvn clean test -Pspark-1.6

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 1.6, Please check CI http://144.76.159.231:8080/job/ApacheCarbonPRBuilder/179/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder/2759/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 1.6, Please check CI http://144.76.159.231:8080/job/ApacheCarbonPRBuilder/180/

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/carbondata-pr-spark-1.6/688/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder/2789/

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/carbondata-pr-spark-1.6/724/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 1.6, Please check CI http://144.76.159.231:8080/job/ApacheCarbonPRBuilder/210/

File: core/src/main/java/org/apache/carbondata/core/metadata/schema/partition/RangeIntervalPartition.java [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.carbondata.core.metadata.schema.partition;
+
+
+public class RangeIntervalPartition extends AbstractPartition {
```
Comment (lionelcao) [inline core/src/main/java/org/apache/carbondata/core/metadata/schema/partition/RangeIntervalPartition.java]: AbstractPartition has been removed. Now only store rangeInfo/ListInfo in partitionInfo.

File: core/src/main/java/org/apache/carbondata/core/scan/filter/partition/PartitionFilterUtil.java [hunk 2]
```diff
@@ -264,4 +265,81 @@ public static BitSet getPartitionMapForRangeFilter(PartitionInfo partitionInfo,
...
+    for (; partitionIndex < numPartitions; partitionIndex++) {
+      result = comparator.compare(filterValue, PartitionUtil.getDataBasedOnDataType(
+          values.get(partitionIndex), partitionColumnDataType, timestampFormatter, dateFormatter));
+      if (result <= 0) {
+        break;
+      }
+    }
+    if (partitionIndex == numPartitions) {
+      // filter value is not in given partition, should get partitionId dynamically
+      int partitionId = partitioner.getPartition(filterValue);
```
Comment (lionelcao) [inline core/src/main/java/org/apache/carbondata/core/scan/filter/partition/PartitionFilterUtil.java]: this should be partitionIndex

File: core/src/main/java/org/apache/carbondata/core/scan/filter/partition/PartitionFilterUtil.java [hunk 3]
```diff
@@ -264,4 +265,81 @@ public static BitSet getPartitionMapForRangeFilter(PartitionInfo partitionInfo,
...
+      partitionMap.set(0);
+      if (isGreaterThan) {
+        // GreaterThan(>), GreaterThanEqualTo(>=)
+        partitionMap.set(partitionId, partitioner.numPartitions());
```
Comment (lionelcao) [inline core/src/main/java/org/apache/carbondata/core/scan/filter/partition/PartitionFilterUtil.java]: this should be partition index

File: core/src/main/java/org/apache/carbondata/core/scan/partition/RangeIntervalPartitioner.java [hunk 4]
```diff
@@ -0,0 +1,206 @@
...
+    while (true) {
+      lastCalendar.add(Calendar.MONTH, 1);
+      boundsList.add(boundsList.size(), lastCalendar.getTimeInMillis());
+      ++numPartitions;
+      if (keyCal.compareTo(lastCalendar) == -1) {
+        partitionId = numPartitions;
+        break;
+      }
+    }
+    return partitionId;
```
Comment (lionelcao) [inline core/src/main/java/org/apache/carbondata/core/scan/partition/RangeIntervalPartitioner.java]: this should be partition index

File: core/src/main/java/org/apache/carbondata/core/scan/partition/RangeIntervalPartitioner.java [hunk 5]
```diff
@@ -0,0 +1,206 @@
...
+    while (true) {
+      lastCalendar.add(Calendar.YEAR, 1);
+      boundsList.add(boundsList.size(), lastCalendar.getTimeInMillis());
+      ++numPartitions;
+      if (keyCal.compareTo(lastCalendar) == -1) {
+        partitionId = numPartitions;
+        break;
+      }
+    }
+    return partitionId;
```
Comment (lionelcao) [inline core/src/main/java/org/apache/carbondata/core/scan/partition/RangeIntervalPartitioner.java]: this should be partition index

File: core/src/main/java/org/apache/carbondata/core/scan/partition/RangeIntervalPartitioner.java [hunk 6]
```diff
@@ -0,0 +1,206 @@
...
+    while (true) {
+      lastCalendar.add(Calendar.WEEK_OF_MONTH, 1);
+      boundsList.add(boundsList.size(), lastCalendar.getTimeInMillis());
+      ++numPartitions;
+      if (keyCal.compareTo(lastCalendar) == -1) {
+        partitionId = numPartitions;
+        break;
+      }
+    }
+    return partitionId;
```
Comment (lionelcao) [inline core/src/main/java/org/apache/carbondata/core/scan/partition/RangeIntervalPartitioner.java]: this should be partition index

File: core/src/main/java/org/apache/carbondata/core/scan/partition/RangeIntervalPartitioner.java [hunk 7]
```diff
@@ -0,0 +1,206 @@
...
+    while (true) {
+      lastCalendar.add(Calendar.DAY_OF_WEEK, 1);
+      boundsList.add(boundsList.size(), lastCalendar.getTimeInMillis());
+      ++numPartitions;
+      if (keyCal.compareTo(lastCalendar) == -1) {
+        partitionId = numPartitions;
+        break;
+      }
+    }
+    return partitionId;
```
Comment (lionelcao) [inline core/src/main/java/org/apache/carbondata/core/scan/partition/RangeIntervalPartitioner.java]: this should be partition index

File: core/src/main/java/org/apache/carbondata/core/scan/partition/RangeIntervalPartitioner.java [hunk 8]
```diff
@@ -0,0 +1,206 @@
...
+    while (true) {
+      lastCalendar.add(Calendar.HOUR_OF_DAY, 1);
+      boundsList.add(boundsList.size(), lastCalendar.getTimeInMillis());
+      ++numPartitions;
+      if (keyCal.compareTo(lastCalendar) == -1) {
+        partitionId = numPartitions;
+        break;
+      }
+    }
+    return partitionId;
```
Comment (lionelcao) [inline core/src/main/java/org/apache/carbondata/core/scan/partition/RangeIntervalPartitioner.java]: this should be partition index

File: integration/spark-common/src/main/scala/org/apache/carbondata/spark/util/CommonUtil.scala [hunk 9]
```diff
@@ -188,7 +189,7 @@ object CommonUtil {
           rangeInfo.get.split(",").map(_.trim).foreach(
             isValid &= validateTypeConvert(partitionerFields(0), _))
         }
-        case "RANGE_INTERVAL" => isValid = false
+        case "RANGE_INTERVAL" => isValid = rangeIntervalInfo.isDefined
```
Comment (lionelcao) [inline integration/spark-common/src/main/scala/org/apache/carbondata/spark/util/CommonUtil.scala]: Is there any validate function for range interval info?

File: core/src/main/java/org/apache/carbondata/core/scan/partition/RangeIntervalPartitioner.java (hunk 8 above)
Comment (lionelcao) [inline core/src/main/java/org/apache/carbondata/core/scan/partition/RangeIntervalPartitioner.java]: this should be partition index

File: core/src/main/java/org/apache/carbondata/core/scan/partition/RangeIntervalPartitioner.java (hunk 7 above)
Comment (lionelcao) [inline core/src/main/java/org/apache/carbondata/core/scan/partition/RangeIntervalPartitioner.java]: this should be partition index

File: core/src/main/java/org/apache/carbondata/core/scan/partition/RangeIntervalPartitioner.java (hunk 6 above)
Comment (lionelcao) [inline core/src/main/java/org/apache/carbondata/core/scan/partition/RangeIntervalPartitioner.java]: this should be partition index

File: core/src/main/java/org/apache/carbondata/core/scan/partition/RangeIntervalPartitioner.java (hunk 4 above)
Comment (lionelcao) [inline core/src/main/java/org/apache/carbondata/core/scan/partition/RangeIntervalPartitioner.java]: this should be partition index

File: core/src/main/java/org/apache/carbondata/core/scan/partition/RangeIntervalPartitioner.java (hunk 5 above)
Comment (lionelcao) [inline core/src/main/java/org/apache/carbondata/core/scan/partition/RangeIntervalPartitioner.java]: this should be partition index

File: core/src/main/java/org/apache/carbondata/core/scan/filter/partition/PartitionFilterUtil.java (hunk 2 above)
Comment (lionelcao) [inline core/src/main/java/org/apache/carbondata/core/scan/filter/partition/PartitionFilterUtil.java]: this should be partition index

**Comment by CarbonDataQA** [comment]
Can one of the admins verify this patch?

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/44/

**Comment by ravipesala** [comment]
SDV Build Fail , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/1221/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/590/

**Pull request closed without merging**